cladecombiner/aggregator.py

```diff
@@ -0,0 +1,228 @@
+            a target will be mapped to Taxon("other"). If False, any such taxa
+            will be mapped to themselves.
+        """
+        unknown_targets = [
+            taxon
+            for taxon in targets
+            if not taxonomy_scheme.is_valid_taxon(taxon)
+        ]
+        if len(unknown_targets) > 0:
+            raise ValueError(
+                "The following targets are not valid taxa according to the provided taxonomy scheme: "
+                + str(unknown_targets)
+            )
+        self.targets = [taxon for taxon in targets]
+        self.taxonomy_scheme = taxonomy_scheme
+        self.agg_other = unmapped_are_other
+        self.warn = warn
+        if sort_clades:
+            self.targets = sort_taxa(self.targets, self.taxonomy_scheme)
+
+    def _validate_inputs(self, input_taxa: Iterable[Taxon]) -> None:
+        unknown_taxa = [
+            taxon
+            for taxon in input_taxa
+            if not self.taxonomy_scheme.is_valid_taxon(taxon)
+        ]
+        if len(unknown_taxa) > 0:
+            raise ValueError(
+                "The following targets are not valid taxa according to the provided taxonomy scheme: "
+                + str(unknown_taxa)
+            )
+
+    def _check_missing(self, agg_map: dict[Taxon, Taxon]):
+        if self.warn:
+            used_targets = set(agg_map.values())
+            unused_targets = [
+                target for target in self.targets if target not in used_targets
+            ]
+            if len(unused_targets) > 0:
+                warn(
+                    f"The aggregation does not make use of the following input targets: {unused_targets}."
+                )
+
+    def aggregate(self, input_taxa: Iterable[Taxon]) -> Aggregation:
+        self._validate_inputs(input_taxa)
+        agg_map: dict[Taxon, Taxon] = {}
+        stack = set(input_taxa)
+        for target in self.targets:
+            children = self.taxonomy_scheme.descendants(target, True)
+            sub_map = {taxon: target for taxon in stack if taxon in children}
+            agg_map = agg_map | sub_map
+            stack.difference_update(set(agg_map.keys()))
+
+        if len(stack) > 0:
+            if self.agg_other:
+                cleanup = HomogenousAggregator(
+                    Taxon("other", False)
+                ).aggregate(stack)
+            else:
+                cleanup = SelfAggregator().aggregate(stack)
+            agg_map = agg_map | cleanup
+
+        self._check_missing(agg_map)
+
+        return Aggregation(input_taxa, agg_map)
+
+
+class HomogenousAggregator(Aggregator):
+    """
+    Aggregation of every taxon to some catch-all taxon.
+    """
+
+    def __init__(self, taxon: Taxon):
+        self.agg_taxon = taxon
+
+    def aggregate(self, input_taxa: Iterable[Taxon]) -> Aggregation:
+        return Aggregation(
+            input_taxa, {taxon: self.agg_taxon for taxon in input_taxa}
+        )
+
+
+class SelfAggregator(Aggregator):
```

Comment:
I kind of liked "trivial"

Reply:
I wanted to be more descriptive. Pretty much any of the aggregators implemented right now have claim to being "trivial" in some sense or another.